### DISCUSSION TITLE
LuaLS settings set to empty table?

### DISCUSSION THREAD

**Discussion by michelesr** [folke/lazydev.nvim#33] [Q&A] [answered]
Hello. I'm running Lazydev on Nvim 0.10 on MacOS ... contrary to Linux, where it works fine, when inspecting the luals settings with `:lua print(vim.inspect(vim.lsp.get_active_clients()))`, the `settings` table is empty.

`:Lazy debug` shows this:

```
## ~/.config/nvim
- **runtime** `/opt/homebrew/Cellar/neovim/0.10.0/share/nvim/runtime/lua`
- **nvim** `~/.config/nvim/lua`
- **cmp-nvim-lsp** `~/.local/share/nvim/lazy/cmp-nvim-lsp/lua`
- **mason-lspconfig.nvim** `~/.local/share/nvim/lazy/mason-lspconfig.nvim/lua`
- **nvim-lspconfig** `~/.local/share/nvim/lazy/nvim-lspconfig/lua`
- **yaml-companion.nvim** `~/.local/share/nvim/lazy/yaml-companion.nvim/lua`
```lua
settings = {
  Lua = {
    runtime = {
      path = { "?.lua", "?/init.lua" },
      pathStrict = true,
      version = "LuaJIT"
    },
    workspace = {
      checkThirdParty = false,
      ignoreDir = { "/lua" },
      library = { "/opt/homebrew/Cellar/neovim/0.10.0/share/nvim/runtime/lua", "/Users/xxx/.config/nvim/lua", "/Users/xxx/.local/share/nvim/lazy/cmp-nvim-lsp/lua", "/Users/xxx
i/.local/share/nvim/lazy/mason-lspconfig.nvim/lua", "/Users/xxx/.local/share/nvim/lazy/nvim-lspconfig/lua", "/Users/xxx/.local/share/nvim/lazy/yaml-companion.nvim/lua" }
    }
  },
  <metatable> = {
    __tostring = <function 1>
  }
}
```

With neodev.nvim, the runtime path was populated correctly. Can someone suggest how to troubleshoot this?

**Comment by michelesr** [comment]
Interestingly, it sets stuff correctly when I set the cd to `~/.local/share/nvim/lazy/lazydev.nvim` and inspect the `lua/lazydev/init.lua` file.

**Comment by folke** [comment, marked as answer]
Can you at least tell me what is not working?
The settings are per workspace, so they might look empty, but they aren't.
lazydev implements scoped workspace comfiguration

**Reply by michelesr** [reply]
Sure. When I open something from my nvim config, there's nothing in the libraries for the language server settings, and so things like `vim` are unknown. It works just fine when I set my pwd to the lazydev source root and open a lua file from there. I'm playing around to see what's happening, but I have no clue so far, the only think I can see is that lazydev is doing the table extend correctly, but then when I run `:lua print(vim.inspect(vim.lsp.get_client_by_id(1).settings))` it prints `vim.empty_dict()` (I've checked the id that is correct, and I've used also the method to get all clients, and that's the only one)... it might not actually be a lazydev issue after all, I'll try to disable other plugins to see if I find what's happening.

**Reply by michelesr** [reply]
@folke do you happen to have, by any chance, a minimal init lua file for neovim that I can use to test a fresh setup of lazydev with lua language server?

**Comment by michelesr** [comment]
On my nvim config:

<img width="836" alt="Screenshot 2024-06-14 at 11 43 40" src="https://github.com/folke/lazydev.nvim/assets/2516077/679aa7d5-be18-4917-ba49-c6469592a812">

On lazydev source:
<img width="816" alt="Screenshot 2024-06-14 at 11 44 02" src="https://github.com/folke/lazydev.nvim/assets/2516077/e0a845f0-bacd-4ff1-8d7a-b523adf825f2">

**Comment by michelesr** [comment]
> The settings are per workspace, so they might look empty, but they aren't.

Gonna close since this is true, the settings seen when you get the client from lua are a empty table even when the `vim` global is resolved correctly. I'm still not sure what's going on, but it can't load properly for me in the nvim config directory, and it works fine in the lazydev directory, any advice on the matter would be appreciated. Thanks!

**Reply by folke** [reply]
In that lua file, where it fails, what does `:=vim.lsp.buf.list_workspace_folders()` show?
And same for `:LspInfo`?
Just to be sure, that undefined is not coming from selene or luacheck or anything?

**Reply by michelesr** [reply]
> In that lua file, where it fails, what does `:=vim.lsp.buf.list_workspace_folders()` show?

```
{ "/Users/xxx/.config/nvim" }
```

> And same for `:LspInfo`

```
 Press q or <Esc> to close this window. Press <Tab> to view server doc.
 
 Language client log: /Users/xxx/.local/state/nvim/lsp.log
 Detected filetype:   lua
 
 1 client(s) attached to this buffer: 
 
 Client: lua_ls (id: 1, bufnr: [1])
 	filetypes:       lua
 	autostart:       true
 	root directory:  /Users/xxx/.config/nvim
 	cmd:             /Users/xxx/.local/share/nvim/mason/bin/lua-language-server
 
 Configured servers list: ruff, pyright, yamlls, lua_ls
```

> Just to be sure, that undefined is not coming from selene or luacheck or anything?|

No, I don't use it. 

It's really weird, I have the same neovim setup (version, and lazy lock file) on Linux and it works fine there, I really don't know what's causing this on MacOS, but neodev.nvim works fine there.

**Reply by michelesr** [reply]
This is from `:LazyDev` in the same buffer.

```
## ~/.config/nvim
- **runtime** `/opt/homebrew/Cellar/neovim/0.10.0/share/nvim/runtime/lua`
- **nvim** `~/.config/nvim/lua`
- **cmp-nvim-lsp** `~/.local/share/nvim/lazy/cmp-nvim-lsp/lua`
- **mason-lspconfig.nvim** `~/.local/share/nvim/lazy/mason-lspconfig.nvim/lua`
- **nvim-lspconfig** `~/.local/share/nvim/lazy/nvim-lspconfig/lua`
- **yaml-companion.nvim** `~/.local/share/nvim/lazy/yaml-companion.nvim/lua`
```lua
settings = {
  Lua = {
    runtime = {
      path = { "?.lua", "?/init.lua" },
      pathStrict = true,
      version = "LuaJIT"
    },
    workspace = {
      checkThirdParty = false,
      ignoreDir = { "/lua" },
      library = { "/opt/homebrew/Cellar/neovim/0.10.0/share/nvim/runtime/lua", "/Users/xxx/.config/nvim/lua", "/Users/xxx/.local/share/nvim/lazy/cmp-nvim-lsp/lua", "/Users/xxx/.local/share/nvim/lazy/mason-lspconfig.nvim/lua", "/Users/xxx/.local/share/nvim/lazy/nvim-lspconfig/lua", "/Users/xxx/.local/share/nvim/lazy/yaml-companion.nvim/lua" }
    }
  },
  <metatable> = {
    __tostring = <function 1>
  }
}
```

**Reply by folke** [reply]
What file exactly does it fail in? Is it your `init.lua`?
After LuaLS loads, can you make an edit to the file and see whether the diagnostics disappear?
Sometimes LuaLS or Neovim seem to show oudated diagnostics.

Edit: all your outputs look fine btw

**Reply by michelesr** [reply]
> What file exactly does it fail in? Is it your `init.lua`? 

Any file in the nvim config folder and subdirectories, I don't have a `init.lua` in the root, but I have a lua folder with different files that I use as my config and source from `init.vim`. Creating a new file in that folder, e.g. `a.lua` yields the same result; if I create a `a.lua` file e.g. in `/tmp/foo` I can resolve `vim` fine there.

> After LuaLS loads, can you make an edit to the file and see whether the diagnostics disappear? Sometimes LuaLS or Neovim seem to show oudated diagnostics.

Tried that, editing, exiting insert mode, saving, and reloading, and didn't help.

**Reply by folke** [reply]
are you using lspconfig?

Because the only thing that;s a bit odd is that you only have that root dir once in the workspace folders.

There's actually a buf in lspconfig that dupicates the first workspace folder you open, so the below should have two identical folders

```
{ "/Users/xxx/.config/nvim" }
```

**Reply by folke** [reply]
Does it resolve other modules correctly?
Like try adding `local cmp = require("cmp")` in a file that doesnt work. 
Is cmp resolved correctly?

**Reply by michelesr** [reply]
> are you using lspconfig?

Yes, I believe so. https://github.com/michelesr/nvim-config/blob/99f23f5ef76800fce83d61c2fe302f9a1fedfb4d/lua/config/lspconfig.lua#L145

**Reply by folke** [reply]
Is that /opt/homebrew/Cellar/neovim/0.10.0/share/nvim/runtime/lua NEovim version the latest stable and not an old nightly?

Is your `$HOME` correct?

**Reply by michelesr** [reply]
```
NVIM v0.10.0
Build type: Release
LuaJIT 2.1.1716656478
Run "nvim -V1 -v" for more info
```

`:= os.getenv('HOME')` returns the correct path

**Reply by michelesr** [reply]
> Does it resolve other modules correctly? Like try adding `local cmp = require("cmp")` in a file that doesnt work. Is cmp resolved correctly?

No, but it does work fine e.g. in `~/a.lua`.

**Reply by michelesr** [reply]
Not sure if it helps... but is there a way to actually determine the final workspace configuration that is applied to the lua language server? Is that stored in some vim specific table, or does lazydev totally manages that independently? I could try to print those table to determine if there's something wrong that it not detected by lazydev directly... just a guess, probably I'm talking nonsense :-)

**Reply by folke** [reply]
Doing `:LazyDev` in the problematic files, prints exactly that

**Reply by folke** [reply]
This might be your problem for some reason: https://github.com/folke/lazydev.nvim/blob/7cbb524c85f87017df9c1ea2377a1d840ad8ed51/lua/lazydev/lsp.lua#L35

Try removing that return and see if it works then.

Also, are you sure you're on a recent LuaLS version?

**Reply by michelesr** [reply]
> Try removing that return and see if it works then.

No, it doesn't work, unfortunately.

> Also, are you sure you're on a recent LuaLS version?

3.9.3

**Reply by michelesr** [reply]
I made a screenshot with Mac and a Linux docker container side by side, for comparison:

<img width="1728" alt="Screenshot 2024-06-15 at 00 12 10" src="https://github.com/folke/lazydev.nvim/assets/2516077/75d396b7-33a7-400c-94cb-60d7bce28e6b">

**Reply by michelesr** [reply]
I've also tried to delete `~/.cache/nvim` and `~/.local/share/nvim` but it didn't help either. Also tried with nightly and no luck.

**Reply by michelesr** [reply]
@folke I'm starting to think it could be something related to lua-language-server, I grepped the debug lsp log to find the did change configuration request and found this:

```
[DEBUG][2024-06-15 09:49:23] .../vim/lsp/rpc.lua:286    "rpc.send"      {  jsonrpc = "2.0",  method = "workspace/didChangeConfiguration",  params = {    settings = {      Lua = {        runtime = {          path = { "?.lua", "?/init.lua" },          pathStrict = true,          version = "LuaJIT"        },        workspace = {          checkThirdParty = false,          library = { "/opt/homebrew/Cellar/neovim/0.10.0/share/nvim/runtime/lua", "/Users/michelesorcinelli/.local/share/nvim/lazy/cmp-nvim-lsp/lua", "/Users/michelesorcinelli/.local/share/nvim/lazy/mason-lspconfig.nvim/lua", "/Users/michelesorcinelli/.local/share/nvim/lazy/nvim-lspconfig/lua", "/Users/michelesorcinelli/.local/share/nvim/lazy/yaml-companion.nvim/lua" }        }      },      <metatable> = {        __tostring = <function 1>      }    }  }}
[DEBUG][2024-06-15 09:49:23] .../vim/lsp/rpc.lua:408    "rpc.receive"   {  id = 7,  jsonrpc = "2.0",  method = "workspace/configuration",  params = {    items = { {        scopeUri = "file:///Users/michelesorcinelli/.config/nvim",        section = "Lua"      }, {        scopeUri = "file:///Users/michelesorcinelli/.config/nvim",        section = "files.associations"      }, {        scopeUri = "file:///Users/michelesorcinelli/.config/nvim",        section = "files.exclude"      }, {        scopeUri = "file:///Users/michelesorcinelli/.config/nvim",        section = "editor.semanticHighlighting.enabled"      }, {        scopeUri = "file:///Users/michelesorcinelli/.config/nvim",        section = "editor.acceptSuggestionOnEnter"      } }  }}
```

do you notice anything wrong here?

**Reply by folke** [reply]
That looks normal

**Comment by michelesr** [comment]
I've figured out what the problem was, there was a `.luarc.json` file in the nvim config folder (and it was in `.gitignore` so it took me a while to notice it) that was referencing some neodev types libraries and other uv things: getting rid of the file fixed the issue. 

@folke thanks for the support, I really appreciate you helping me even if you had no way to reproduce.

**Reply by folke** [reply]
Great!
I just added a warning to `:LazyDev` in case a `.luarc.json` is found
